### PR TITLE
fix Issue 12486 - Function returning struct isn't called if `enum` of its result is accessed

### PR DESCRIPTION
**Pull request by FeepingCreature**
Fix issue 12486

When accessing an enum member of a struct returned by an expression with side effects, don't discard the expression but evaluate it anyways. (No, this is not safe even if the expression is pure: it might loop infinitely.)

### CODE REVIEW

**Comment by FeepingCreature** [comment]
Ah, I see, it has to be an expression but it is a struct.
Wait, huh?

**Comment by FeepingCreature** [comment]
Is there a way to say "the expression is trivial"? I don't like enumerating all the different kinds of trivial expressions manually. Maybe optimize just the comma expression, somehow?

**Comment by FeepingCreature** [comment]
Weird, auto-tester seems to be running the wrong version of the code. Gonna wait for everything to finish, then kick it off again.

**Comment by wilzbach** [comment]
> Weird, auto-tester seems to be running the wrong version of the code. Gonna wait for everything to finish, then kick it off again.

No need to wait. auto-tester automatically refreshes itself though it might take a bit until it reruns your PR as there's usually a lot of traffic.

**Comment by FeepingCreature** [comment]
Okay. `e.optimize` is _definitely_ the wrong way to go about this. Is there any way to decide whether an Expression is relatively trivial?

**Comment by FeepingCreature** [comment]
_:stares at ci Jenkins:_

Do you ever get that feeling that, while all the tests are green, they really *shouldn't* be?

**Comment by wilzbach** [comment]
> Do you ever get that feeling that, while all the tests are green, they really shouldn't be?

Jenkins doesn't run the testsuite. It just compiles dmd and then builds ~40 projects and runs their testsuite.

**Comment by FeepingCreature** [comment]
:stares at AppVeyor:

Do you ever get that feeling that, while all the tests are green, they really shouldn't be?

**Comment by FeepingCreature** [comment]
Is it stuck? auto-tester has been at pending: 8 for hours now.

**Comment by wilzbach** [comment]
> Is it stuck? auto-tester has been at pending: 8 for hours now.

No, auto-tester deletes all builds whenever a new PR is merged at DMD, druntime or phobos. Click on the link to see the history.
Don't worry too much about it. Once a PR is approved, it gets priority and lands in a special auto-merge queue.

**Comment by FeepingCreature** [comment]
How do I find a reviewer?

File: src/dmd/typesem.d
Comment (FeepingCreature) [inline src/dmd/typesem.d]: The point here is to avoid being forced to optimize the expression if we know the lhs of the comma expression would be trivial, since optimization may change the meaning but is sometimes necessary so we don't have a pointless comma expression lying around. I wish there was a better way to do this.

**Comment by ibuclaw** [comment]
> How do I find a reviewer?

Ping or assign me and I'll get round to it. 😉

**Comment by FeepingCreature** [comment]
Um ... @ibuclaw ping? I don't know how to assign people.

**Comment by FeepingCreature** [comment]
I debugged what's happening with the Phobos tests.

The Cartesian products being computed in std.setops are ridiculously too large. They *look* small, but due to the fact that 3D nested Cartesian product is strongly biased towards Z spread, actually take billions to trillions of cycles to run.

This was not noticed before because due to the bug, they essentially passed by coincidence. (`canFind` calls `any`.)

File: src/dmd/typesem.d
Comment (RazvanN7) [reply src/dmd/typesem.d]: Please add this explanation as a comment in the code. It will be useful in the future when people will try to understand the code.

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/typesem.d
Comment (FeepingCreature) [reply src/dmd/typesem.d]: Documentation added.

File: src/dmd/typesem.d
Comment (WalterBright) [inline src/dmd/typesem.d]: See https://github.com/dlang/dmd/blob/master/src/dmd/sideeffect.d

File: src/dmd/typesem.d
Comment (WalterBright) [inline src/dmd/typesem.d]: This fix looks like it's in the wrong place. `expressionSemantic()` should *not* be removing expressions with side effects. Putting the fix here is papering over the problem for this case, and it'll just cause other problems. Need to find the root problem.

**Review by WalterBright — CHANGES_REQUESTED** [review]
Fix is in the wrong place.

File: src/dmd/typesem.d
Comment (FeepingCreature) [reply src/dmd/typesem.d]: Right, the problem is if I keep the side effects via comma expression, as I have to in order to get correct code in the case call().ENUM, then I end up with unusable comma expressions downstream. I can't fix it earlier, because this is where the access resolves.

Comment (WalterBright) [reply src/dmd/typesem.d]: > what DMD does right now is removing side effects from expressions in that code: call().ENUM becomes ENUM, with call() being discarded.

That's a bug, as call() has side effects. That's where it needs to be fixed.

> The problem is fixing that without breaking enum access for trivial lhs, say alias foo = this.bar.

Maybe that's there the special case should go.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: I don't understand what you mean. This is where that happens.

File: src/dmd/typesem.d
Comment (FeepingCreature) [reply src/dmd/typesem.d]: As I understand it, a function that never returns but is pure will still be considered as not having side effects by this code? But we must not optimize it away: we may optimize >1 calls to 1 call, but we mustn't optimize 1 call to 0 calls.

**Comment by FeepingCreature** [comment]
Ping?

**Comment by quickfur** [comment]
@FeepingCreature  I think what Walter meant is that the proper fix should probably be somewhere in `sideeffect.d` so that the code here, unchanged, will do the right thing.  In other words, the fact that the code here is discarding side effects when it shouldn't, means that `sideeffect.d` (or whatever else is the root cause of this problem) isn't doing its job properly.

**Comment by quickfur** [comment]
More specifically, there may be a defect somewhere in `hasSideEffect` or `callSideEffect` that isn't detecting the side-effect when it should be.  Or perhaps `hasSideEffect` or `callSideEffect` isn't getting called when it should be, resulting in the compiler throwing away code thinking that there are no side-effects, when in fact there are.

**Comment by FeepingCreature** [comment]
Well, are infinite loops *supposed* to count as side effects? If not, istm there is a lack of an idiom here.

**Comment by ibuclaw** [comment]
> Well, are infinite loops supposed to count as side effects? If not, istm there is a lack of an idiom here. isDisposable or such.

All calls should be considered as a side effect until inlined. There is `isTrivialExp` in the sideeffects module that likely fits your needs.

**Comment by FeepingCreature** [comment]
Thanks, that seems sensible.

I still don't get why Walter thinks the fix is in the wrong place. This is the place that throws away the side effect right now.

**Comment by FeepingCreature** [comment]
@WalterBright I tried to keep it as a DotExp instead of forcing it into a CommaExp, but if I do so all the way through codegen, it produces invalid code because it seems to try and treat the enum access as a struct read. ISTM that CommaExp is just how enum dot accesses to expressions with side effects have to look right now. If it's possible to make a better fix for this, I don't know how. I don't think I can get the code materially better than this.

**Comment by FeepingCreature** [comment]
ping

**Review by ibuclaw — APPROVED** [review]
Seems reasonable to me.

@FeepingCreature - please rebase to restart failing tests.
@WalterBright - check this again.

**Comment by wilzbach** [comment]
> please rebase to restart failing tests.

No need. auto-tester will automatically restart itself and all other CIs are passing.
Regarding the failures on auto-tester - they are spurious and I just submitted:

https://github.com/dlang/phobos/pull/6347

which hopefully helps to weed them out.

**Comment by FeepingCreature** [comment]
ping

**Comment by ibuclaw** [comment]
I'll compile this patch in locally and inspect the change done, then will try to break it. 😉

If all is OK I see no problem with committing this.

File: src/dmd/typesem.d
Comment (ibuclaw) [inline src/dmd/typesem.d]: @FeepingCreature - you could also use `hasSideEffects(e)` (or `e.hasSideEffects()` if you prefer UFCS).  When having an initial glance at the implementation, it didn't seem like it was a strong enough guarantee, but having a closer look, only strongly pure functions would have their calls omitted here.  Which is harmless.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: As far as I can tell, `pure void foo() { while (true) { } }` is strongly pure, but must under no circumstances be optimized away.

Comment (ibuclaw) [reply src/dmd/typesem.d]: Ah, bugger those looping pure functions. :-)

Yep, makes sense, and LGTM.

**Comment by ibuclaw** [comment]
@WalterBright - Any last comments before I dismiss your stale review?

File: src/dmd/typesem.d
Comment (ibuclaw) [reply src/dmd/typesem.d]: @FeepingCreature - For brownie points, you can add a comment with a link to the bugzilla reference.

**Pull request merged**
